kong/tools/wrpc/init.lua

```diff
@@ -170,382 +71,93 @@ function wrpc_peer:receive()
+--- @param arg(table) arguments of the call, like {config = config}
+function _M:call_no_return(name, arg)
+  local future_to_wait = self:call(name, arg)
+
+  return future_to_wait:drop()
 end
 
+
 --- encodes and sends a wRPC message.
 --- Assumes protocol fields are already filled (except `.seq` and `.deadline`)
 --- and payload data (if any) is already encoded with the right type.
 --- Keeps track of the sequence number and assigns deadline.
-function wrpc_peer:send_payload(payload)
+function _M:send_payload(payload)
   local seq = self.seq
   payload.seq = seq
   self.seq = seq + 1
 
+  -- protobuf may confuse with 0 value and nil(undefined) under some set up
+  -- so we will just handle 0 as nil
   if not payload.ack or payload.ack == 0 then
-    payload.deadline = ngx.now() + DEFAULT_EXPIRATION_DELAY
+    payload.deadline = ngx_now() + DEFAULT_EXPIRATION_DELAY
   end
 
-  self:send(self.encode("wrpc.WebsocketPayload", {
+  self:send(pb_encode("wrpc.WebsocketPayload", {
     version = "PAYLOAD_VERSION_V1",
     payload = payload,
   }))
 end
 
-function wrpc_peer:send_remote_payload(msg, name)
-  local payload = self.decode("wrpc.PayloadV1", msg)
-  self.response_queue[self.seq] = remote_waiter(self.channel_dict, name)
-  return self:send_payload(payload)
-end
-
---- Handle RPC data (mtype == MESSAGE_TYPE_RPC).
---- Could be an incoming method call or the response to a previous one.
---- @param payload table decoded payload field from incoming `wrpc.WebsocketPayload` message
-function wrpc_peer:handle(payload)
-  local rpc = self.service:get_rpc(payload.svc_id .. '.' .. payload.rpc_id)
-  if not rpc then
-    self:send_payload({
-      mtype = "MESSAGE_TYPE_ERROR",
-      error = {
-        etype = "ERROR_TYPE_INVALID_SERVICE",
-        description = "Invalid service (or rpc)",
-      },
-      svc_id = payload.svc_id,
-      rpc_id = payload.rpc_id,
-      ack = payload.seq,
-    })
-    return nil, "INVALID_SERVICE"
-  end
-
-  local ack = tonumber(payload.ack) or 0
-  if ack > 0 then
-    -- response to a previous call
-    local response_waiter = self.response_queue[ack]
-    if response_waiter then
-
-      if response_waiter.deadline and response_waiter.deadline < ngx.now() then
-        if response_waiter.expire then
-          response_waiter:expire()
-        end
-
-      else
-        if response_waiter.raw then
-          response_waiter:handle(payload)
-        else
-          response_waiter:handle(decodearray(self.decode, rpc.output_type, payload.payloads))
-        end
-      end
-      self.response_queue[ack] = nil
-
-    elseif rpc.response_handler then
-      pcall(rpc.response_handler, self, decodearray(self.decode, rpc.output_type, payload.payloads))
-    end
-
-  else
-    -- incoming method call
-    if rpc.handler then
-      local input_data = decodearray(self.decode, rpc.input_type, payload.payloads)
-      local ok, output_data = ok_wrapper(pcall(rpc.handler, self, table_unpack(input_data, 1, input_data.n)))
```

Comment:
this line is like this to support multiple args.  if that feature is removed, this line should be much simpler